WI #616: Addition of five columns related to `Pricing Currency` (#848)
FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec: specification/attributes/currency_format.md

Reviewer: Would we be looking to add an additional field for currency type in a future specification as we are specifying the type of currency in this document.

Reviewer: This bit is a little confusing some currency related columns could relate to costs so are we specifically meaning the currency related columns which are in the currency format.  Currency-related columns MUST conform to [StringHandling](#stringhandling) requirements when the value is presented in virtual currency (e.g. credits, tokens).
* Currency-related columns MUST be represented as a three-letter alphabetic code as dictated in the governing document [ISO 4217:2015](https://www.iso.org/standard/64758.html) in all other cases.

Reviewer 2: A currency type field may be helpful and would be a good item to open in the backlog if you would like to see one added so it can be considered for future development.

Agreed 'in all other cases' maybe confusing as it makes this normative text not stand on its own and become positional with the previous line. I'll make a suggestion to just say `when the value is presented in national currency (e.g. USD, EUR)`

Author: Thanks for the feedback @jpickardgreen, and thanks for the suggestion @mike-finopsorg!  I've committed that suggestion.  Resolving.

> # Currency Format

Reviewer 3: This, as written, is problematic because it is referenced by Billing currency, which must not be a virtual currency (though we've missed those updates here). It would be a hugely disruptive regression if providers started producing their billing data in a virtual currency. I believe we need to either restructure this to National Currency Format and Currency Format OR somehow reference that only a part of the attribute is applicable for BillingCurrency (we don't have a pattern like this in the spec today, so we would have to define that for this need -- if this is the approach we want to take).

Reviewer 3: I don't think this gives us the safety we are looking for that guarantees it's not in a virtual currency for typical SaaS type scenarios. Remember, most of those are pre-paid. So it's very possible there is only an invoice at the time there was a pre-payment and subsequent covered usage has no additional cost and therefore isn't required to have an invoice. However, these records are 'usage' data that helps track the burn-down of the pre-payment. 

I'm good to resolve if the group believes this isn't needed but I see value in being explicit and not having conflicting requirements:

- BillingCurrency MUST conform to CurrencyCodeFormat requirements.

(which needs updating and should be in this PR since we're renaming CurrencyCodeFormat here)

This suggests: A currency may be one of the following currency types: National currency (e.g. USD, EUR) | Virtual currency (e.g. tokens, credits).

Then the next requirement suggests:

- BillingCurrency MUST match the currency used in the invoice generated by the invoice issuer.

Which in most cases may suggest this only applies to national currencies, but as mentioned above, there's not a strict need for invoices for SaaS for ongoing usage when it's already pre-paid.

Author: First of all, great point that I missed updating the Billing Currency reference to Currency Format.  That's now been addressed in [this commit](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/848/commits/9bac16882992dada52d122a32e6d1e39251a282b).

My original draft of Currency Format was more explicit about the usage of national currency in purchase scenarios, but I changed that per Chris' suggestion in [this resolved thread](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/848/files/2d7b921e3cb3e0cb0782b1d54c20a10cf96b1448#r2002164545). 

If we want to be explicit, then per @ijurica, we could add a normative requirement to Billing Currency such as:

> * BillingCurrency MUST be expressed in national currency (e.g., USD, EUR).

(I do agree with Chris' point from that linked thread that there is indeed the possibility of providers charging in Bitcoin, Ethereum, and the like -- though I can't say I know of a precedent for that today.)

Would that work for you?

Author: I have added that requirement per [this commit](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/848/commits/d78de28594335042ff3ae6044935bffff77337b5).  Resolving.

Columns that contain currency information in cost data following a consistent format reduce friction for FinOps practitioners who consume the data for analysis, reporting, and other use cases.

A currency may be one of the following currency types:

* National currency (e.g. USD, EUR).
* Virtual currency (e.g. tokens, credits).

All columns capturing a currency value, defined in the FOCUS specification, MUST follow the requirements listed below. Custom currency-related columns SHOULD also follow the same formatting requirements.

## Attribute ID

CurrencyFormat

## Attribute Name

Currency Format

## Description

Formatting for currency columns appearing in a [*FOCUS dataset*](#glossary:FOCUS-dataset).

## Requirements

> * Currency-related columns MUST conform to [String Handling](#stringhandling) requirements when the value is presented in virtual currency (e.g. credits, tokens).

Author: Yes, for sure.  Thanks for the callout!

> * Currency-related columns MUST conform to [String Handling](#stringhandling) requirements when [ChargeCategory] is not "Purchase" and the provider presents prices in a virtual currency (e.g. credits, tokens).

Reviewer 4: 1. Does this mean that all purchases MUST be in national currencies?
2. Would refunds also be in national currencies too?
3. In the case of an exchange (like with RI's), is there a case where one virtual currency is exchanged for another and there is a refund of virtual currency A and a purchase of virtual currency B.  Would we then mandate that anything similar must be shown in national currencies?

Author: 1. Purchases: yes.  Do you agree?
2. Refunds: there, I didn't specify.  I don't think we'll know if a refund will be in national or virtual currency, so I left it open.  Do you agree?
3. Exchanges: I don't know, to be honest.  What do you think?

Reviewer 4: 1. I don't know ... it's very feasible that purchases can be made with virtual currencies like Bitcoin.
2. Leaving it open is good, worth mentioning "refund" in some capacity though.
3. I honestly don't know.  Additionally, would a rollover then have to be categorized as a purchase?

The large point is that these 5 columns introduce a lot of complexity across SaaS scenarios that the group isn't as well-versed in as compared to CSPs.  So, I would lean very heavily towards general, over-arching guidance that can cover all the idiosyncracies that we're not aware of.

Author: 1. I think it's a fair point.  I'm open to changing this to be something more general like, "Conform to StringHandling when virtual currency."
2. If I make the change per 1, where we're not putting restrictions on when to state in Pricing or Billing Currency, thankfully that becomes moot.
3. Same answer as 2.

So then, how about:

```suggestion
* Currency-related columns MUST conform to [String Handling](#stringhandling) requirements when the value is presented in virtual currency (e.g. credits, tokens).
```

Author: Cool.  Committing the suggestion.

> * Currency-related columns MUST be represented as a three-letter alphabetic code as dictated in the governing document [ISO 4217:2015](https://www.iso.org/standard/64758.html) in all other cases.

Reviewer 2: ```suggestion
* Currency-related columns MUST be represented as a three-letter alphabetic code as dictated in the governing document [ISO 4217:2015](https://www.iso.org/standard/64758.html) when the value is presented in national currency (e.g. USD, EUR).
```
